bpf/cpu/pyperf.bpf.c

```diff
@@ -0,0 +1,439 @@
+  long unsigned int tls_base = read_tls_base(task);
+  LOG("tls_base 0x%llx", (void *)tls_base);
+
+  GET_OFFSETS();
+
+  // Fetch the thread id.
+
+  // Python 3.11+ uses native_thread_id.
+  if (offsets->py_thread_state.native_thread_id > 0) {
+    u64 thread_id = 0;
+    LOG("offsets->py_thread_state.native_thread_id %d", offsets->py_thread_state.native_thread_id);
+    bpf_probe_read_user(&thread_id, sizeof(thread_id), state->thread_state + offsets->py_thread_state.native_thread_id);
+    LOG("thread_id %d", thread_id);
+    if (thread_id != tid) {
+      LOG("[error] thread_id %d != tid %d", thread_id, tid);
+      goto submit_event;
+    }
+  } else {
+    LOG("offsets->py_thread_state.thread_id %d", offsets->py_thread_state.thread_id);
+    pthread_t pthread_id;
+    bpf_probe_read_user(&pthread_id, sizeof(pthread_id), state->thread_state + offsets->py_thread_state.thread_id);
+    LOG("pthread_id %lu", pthread_id);
+    // 0x10 = offsetof(tcbhead_t, self) for glibc.
+    pthread_t current_pthread_id;
+    bpf_probe_read_user(&current_pthread_id, sizeof(current_pthread_id), (void *)(tls_base + 0x10));
+    LOG("current_pthread_id %lu", current_pthread_id);
+    if (pthread_id != current_pthread_id) {
+      LOG("[error] pthread_id %lu != current_pthread_id %lu", pthread_id, current_pthread_id);
+      goto submit_event;
+    }
+    state->current_pthread = current_pthread_id;
+  }
+
+  // Get pointer to top frame from PyThreadState.
+
+  if (offsets->py_thread_state.frame > -1) {
+    LOG("offsets->py_thread_state.frame %d", offsets->py_thread_state.frame);
+    bpf_probe_read_user(&state->frame_ptr, sizeof(void *), state->thread_state + offsets->py_thread_state.frame);
+  } else {
+    LOG("offsets->py_thread_state.cframe %d", offsets->py_thread_state.cframe);
+    void *cframe;
+    bpf_probe_read_user(&cframe, sizeof(cframe), (void *)(state->thread_state + offsets->py_thread_state.cframe));
+    if (cframe == 0) {
+      LOG("[error] cframe was NULL");
+      goto submit_event;
+    }
+    LOG("cframe 0x%llx", cframe);
+
+    LOG("offsets->py_cframe.current_frame %d", offsets->py_cframe.current_frame);
+    bpf_probe_read_user(&state->frame_ptr, sizeof(state->frame_ptr), (void *)(cframe + offsets->py_cframe.current_frame));
+  }
+  if (state->frame_ptr == 0) {
+    LOG("[error] frame_ptr was NULL");
+    goto submit_event;
+  }
+  LOG("frame_ptr 0x%llx", state->frame_ptr);
+
+  bpf_tail_call(ctx, &programs, PYPERF_STACK_WALKING_PROGRAM_IDX);
+
+submit_event:
+  aggregate_stacks();
+  return 0;
+}
+
+static inline __attribute__((__always_inline__)) u64 get_symbol_id(symbol_t *sym) {
+  int *symbol_id_ptr = bpf_map_lookup_elem(&symbol_table, sym);
+  if (symbol_id_ptr) {
+    return *symbol_id_ptr;
+  }
+
+  u32 zero = 0;
+  u64 *sym_idx = bpf_map_lookup_elem(&symbol_index_storage, &zero);
+  if (sym_idx == NULL) {
+    // Appease the verifier, this will never fail.
+    return 0;
+  }
+
+  u64 idx = __sync_fetch_and_add(sym_idx, 1);
```

Comment:
This atomic increment is most likely the reason why some kernels are not okay with this code. The ID sharding technique from rbperf could be useful here too 😄

Reply:
I'll check it 👍